examples/all-clusters-app/all-clusters-common/src/rvc-modes.cpp

```diff
@@ -0,0 +1,141 @@
+}
+
+void RvcRunModeInstance::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t currentMode = GetCurrentMode();
+
+    // Our business logic states that we can only switch into the mapping state from the idle state.
+    if (NewMode == RvcRunMode::ModeMapping && currentMode != RvcRunMode::ModeIdle)
+    {
+        response.status = static_cast<uint8_t>(ModeBase::StatusCode::kGenericFailure);
+        response.statusText.SetValue(chip::CharSpan("Change to the mapping state is only allowed from idle", 53));
+    }
+
+    response.status = static_cast<uint8_t>(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR RvcRunModeInstance::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex < NumberOfModes())
+    {
+        if (label.size() >= modeOptions[modeIndex].label.size())
+        {
+            CopyCharSpanToMutableCharSpan(modeOptions[modeIndex].label, label);
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR RvcRunModeInstance::GetModeValueByIndex(uint8_t modeIndex, uint8_t & value)
+{
+    if (modeIndex < NumberOfModes())
+    {
+        value = modeOptions[modeIndex].mode;
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR RvcRunModeInstance::GetModeTagsByIndex(uint8_t modeIndex, List<ModeTagStructType> & tags)
+{
+    if (modeIndex < NumberOfModes())
+    {
+        if (tags.size() >= modeOptions[modeIndex].modeTags.size())
+        {
+            std::copy(modeOptions[modeIndex].modeTags.begin(), modeOptions[modeIndex].modeTags.end(), tags.begin());
+            tags.reduce_size(modeOptions[modeIndex].modeTags.size());
+
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR RvcCleanModeInstance::AppInit()
+{
+    return CHIP_NO_ERROR;
+}
+
+void RvcCleanModeInstance::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t rvcRunCurrentMode = GetCurrentMode();
+
+    if (rvcRunCurrentMode == RvcRunMode::ModeCleaning)
+    {
+        response.status = static_cast<uint8_t>(RvcCleanMode::StatusCode::kCleaningInProgress);
+        response.statusText.SetValue(chip::CharSpan("Cannot change the cleaning mode during a clean", 60));
+    }
+
+    response.status = static_cast<uint8_t>(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR RvcCleanModeInstance::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex < NumberOfModes())
+    {
+        if (label.size() >= modeOptions[modeIndex].label.size())
+        {
+            CopyCharSpanToMutableCharSpan(modeOptions[modeIndex].label, label);
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
```

Comment:
All these similar impls used by the different mode select could have been a single derived class of something like "TableBasedModeSelect" given that they are all nearly identical except for HandleChangetoMode. Consider refactoring code to reduce flash usage in a follow-up

Reply:
Added as a todo.